---
title: "BSG HW4"
author: "Sten-Oliver Salumaa, Denys Kovalenko"
date: "13 December 2017"
output: html_document
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(genetics)
```

## Task1 (1p)
#### Myoglobin is an oxygen-binding protein found in muscle tissue. The protein is encoded by the MB gene, which resides on the long arm of chromosome 22. The file MB.rda contains genotype information of unrelated individuals for a set of SNPs in the MB gene. The file contains genotype information in object Y. Load this data into the R environment.

```{r echo=FALSE, warning=FALSE}
load("MB.rda")
```


## Task2 (1p)
####  How many individuals and how many SNPs are there in the database? What percentage of the data is missing? 

```{r echo=FALSE, warning=FALSE}
print("Individual count:")
nrow(Y)
print("SNP count:")
ncol(Y)

total <- ncol(Y)*nrow(Y)
nas <- length(Y[is.na(Y)])
print("Percentage of data missing:")
nas/total*100
```


## Task3 (1p)
#### Assuming all SNPs are bi-allelic, how many haplotypes can theoretically be found for this data set?

```{r echo=FALSE, warning=FALSE}
# not sure but looks like 2**(num of SNPs)
print("Theoretical number of haplotypes:")
2**ncol(Y)
```

## Task4 (2p)
#### Estimate haplotype frequencies using the haplo.stats package (set the minimum posterior probability to 0.001). How many haplotypes do you find? List the haplotypes and their estimated probabilities. Which haplotype is the most common?

```{r echo=FALSE, warning=FALSE}
library(haplo.stats)

# Remove all SNPs with one or more missing observations.
#nmis <- function(x) {
#  y <- sum(is.na(x))
#}

# get the nmis of SNPs
#mispersnp <- apply(Y,1,nmis)
#mispersnp

#dim(Y)
#Y2 <- Y[mispersnp==0,]
#dim(Y2)

Geno2 <- cbind(substr(Y[,1],1,1),substr(Y[,1],2,2))
for(i in 2:ncol(Y)) {
  Geno2 <- cbind(Geno2,substr(Y[,i],1,1),substr(Y[,i],2,2))
}

snpnames <- paste("snp", 1:28,sep="")

out <-  haplo.em(Geno2,
                 locus.label=snpnames,
                 control=haplo.em.control(min.posterior=1e-3))



print("Total haplotypes:")
nrow(out$haplotype)
print("Most popular haplotype:")
paste(tail(out$haplotype,1))
print("Probability of the most popular haplotype:")
tail(out$hap.prob,1)

```


## Task5 (2p)
####  Is the haplotypic constitution of any of the individuals in the database ambiguous or uncertain? If so, for which individuals? What is, in this case, the most likely haplotypic constitution of any possibly uncertain individuals?

```{r echo=FALSE, warning=FALSE}

#function to get heterozygous count per row in data
hetero_count <- function(row){
  row <- row[!is.na(row)]
  res <- substr(row, 1, 1) != substr(row, 2, 2)
  for(i in 2:length(row)){
    append(res, substr(row[i], 1, 1) != substr(row[i], 2, 2))
  }
  return(sum(res))
}


counts <- apply(Y, 1, hetero_count)
ambiguous <- counts[counts > 1]
print("Individuals with ambiguous haplotypes:")
names(ambiguous)

ambi_haplo <- Y[counts > 1, ]
ambi_geno <- cbind(substr(ambi_haplo[,1],1,1),substr(ambi_haplo[,1],2,2))
for(i in 2:ncol(ambi_haplo)) {
  ambi_geno <- cbind(ambi_geno,substr(ambi_haplo[,i],1,1),substr(ambi_haplo[,i],2,2))
}




snpnames2 <- paste("snp", 1:ncol(Y),sep="")
ambi_out <-
  haplo.em(ambi_geno,
  locus.label = snpnames2,
  control = haplo.em.control(min.posterior = 1e-3))

#as.character(haplo$haplotype[which.max(haplo$hap.prob), ])

print("The most likely haplotypic constitution for these ambiguous haplotypes:")
paste(tail(ambi_out$haplotype,1))
print("And its probability:")
tail(ambi_out$hap.prob,1)

```

```{r echo=FALSE, warning=FALSE}


```

## Task6 (1p)
#### Suppose we would delete SNP rs5999890 from the database prior to haplotype estimation. Would this affect the results obtained? Justify your answer. Delete this SNP from the database and estimate again the haplotype frequencies. List the haplotypes and their estimated frequencies.

It shouldn't affect results, cause SNP rs5999890 has only CC values, so it doesn't change possible haplotypes.

```{r echo=FALSE, warning=FALSE}
Y_cut <- Y[,-5]

Geno2_cut <- cbind(substr(Y_cut[,1],1,1),substr(Y_cut[,1],2,2))
for(i in 2:ncol(Y_cut)) {
  Geno2_cut <- cbind(Geno2_cut,substr(Y_cut[,i],1,1),substr(Y_cut[,i],2,2))
}


snpnames <- paste("snp", 1:27,sep="")

out_cut <-  haplo.em(Geno2_cut,
                 locus.label=snpnames,
                 control=haplo.em.control(min.posterior=1e-3))



print("Testing the hypothesis...")
print("Total haplotypes:")
nrow(out_cut$haplotype)
print("Most popular haplotype:")
paste(tail(out_cut$haplotype,1))
print("Probability of the most popular haplotype:")
tail(out_cut$hap.prob,1)
```

The test proves that actually we got same results.

```{r echo=FALSE, warning=FALSE}
print("Listing haplotypes and their estimated frequencies:")
out_cut
```




## Task8 (1p)
#### Simulate a set of independent markers using the the multinomial distribution (R function rmultinom) that mimicks the Myoglobin data in terms of sample size, number of SNPs and minor allele frequencies, assuming HardyWeinberg equilibrium (that is, simulate the markers with multinomial probabilities p2, 2pq and q2, where p is the observed minor allele frequency) Create haplotypes on the basis of the simulated data. Do you find the same number of haplotypes? Can you explain the difference?


```{r echo=FALSE, warning=FALSE, cache=FALSE}


get_allele_freq <- function(x){
  g <- genotype(x[!is.na(x)], sep="")
  s <- summary(g)
  allele_freq <- s$allele.freq[,'Proportion'][1]
  return(c(allele_freq, 1 - allele_freq))
}

allele_freqs <- apply(Y, 2, get_allele_freq)


get_geno_freqs <- function(x){
  homoA <- x[2]**2
  hetAB <- x[1]*x[2]
  homoB <- x[1]**2
  return(c(homoA, hetAB, homoB))
}

geno_freqs <- apply(allele_freqs, 2, get_geno_freqs)


get_genotypes <- function(x){
  probs <- x
  genotype_col <- sample(c("AA", "AB", "BB"), size = nrow(Y), prob = probs, replace=TRUE)
  return(genotype_col)
}


simulated_MBData <- apply(geno_freqs, 2, get_genotypes)

Geno3 <- cbind(substr(simulated_MBData[,1],1,1),substr(simulated_MBData[,1],2,2))
for(i in 2:ncol(simulated_MBData)) {
  Geno3 <- cbind(Geno3,substr(simulated_MBData[,i],1,1),substr(simulated_MBData[,i],2,2))
}


snpnames <- paste("snp", 1:28,sep="")
#length(snpnames)

sim_out <-
  haplo.em(Geno3,
  locus.label = snpnames,
  control = haplo.em.control(min.posterior = 1e-3))



print("The most likely haplotypic constitution for these ambiguous haplotypes (having alleles 'A' and 'B':")
paste(tail(sim_out$haplotype,1))
print("And its probability:")
tail(sim_out$hap.prob,1)
print("Total haplotypes:")
nrow(sim_out$haplotype)

```

The results are very different. Previously under the same settings we found 6 probable haplotypes but now we see close to 130. This suggests that there is an association between different loci of myoglobin dataset (linkage disequilibrium) thus making some haplotypes more probable than others. In the simulated dataset we assumed no linkage (totally independent SNPs) and so we got more uniform distribution of haplotypes and their probabilities.


```{r echo=FALSE, warning=FALSE}



```


```{r echo=FALSE, warning=FALSE}



```


```{r echo=FALSE, warning=FALSE}



```