---
title: "BSG HW3"
author: "Sten-Oliver Salumaa, Denys Kovalenko"
date: "13 December 2017"
output: html_document
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Task1
#### Myoglobin is an oxygen-binding protein found in muscle tissue. The protein is encoded by the MB gene, which resides on the long arm of chromosome 22. The file MB.rda contains genotype information of unrelated individuals for a set of SNPs in the MB gene. The file contains genotype information in object Y. Load this data into the R environment.

```{r echo=FALSE, warning=FALSE}
load("MB.rda")
```


## Task2
####  (1p) How many individuals and how many SNPs are there in the database? What percentage of the data is missing? 

```{r echo=FALSE, warning=FALSE}
print("Individual count:")
nrow(Y)
print("SNP count:")
ncol(Y)

total <- ncol(Y)*nrow(Y)
nas <- length(Y[is.na(Y)])
print("Percentage of data missing:")
nas/total*100
```


## Task3
#### Assuming all SNPs are bi-allelic, how many haplotypes can theoretically be found for this data set?

```{r echo=FALSE, warning=FALSE}
# not sure but looks like 2**(num of SNPs)
print("Theoretical number of haplotypes:")
2**ncol(Y)
```

## Task4
#### Estimate haplotype frequencies using the haplo.stats package (set the minimum posterior probability to 0.001). How many haplotypes do you find? List the haplotypes and their estimated probabilities. Which haplotype is the most common?

```{r echo=FALSE, warning=FALSE}
library(haplo.stats)

# Remove all SNPs with one or more missing observations.
nmis <- function(x) {
  y <- sum(is.na(x))
}

# get the nmis of SNPs
mispersnp <- apply(Y,1,nmis)
#mispersnp

#dim(Y)
Y2 <- Y[mispersnp==0,]
#dim(Y2)
# no se, we have to produce columns like "A G D C A D S D A A" or whatever and thed we can use the other function
# world-class coding right here:
Geno <- cbind(substr(Y[,1],1,1),substr(Y[,1],2,2),
              substr(Y[,2],1,1),substr(Y[,2],2,2),
              substr(Y[,3],1,1),substr(Y[,3],2,2),
              substr(Y[,4],1,1),substr(Y[,4],2,2),
              substr(Y[,5],1,1),substr(Y[,5],2,2),
              substr(Y[,6],1,1),substr(Y[,6],2,2),
              substr(Y[,7],1,1),substr(Y[,7],2,2),
              substr(Y[,8],1,1),substr(Y[,8],2,2),
              substr(Y[,9],1,1),substr(Y[,9],2,2),
              substr(Y[,10],1,1),substr(Y[,10],2,2),
              substr(Y[,11],1,1),substr(Y[,11],2,2),
              substr(Y[,12],1,1),substr(Y[,12],2,2),
              substr(Y[,13],1,1),substr(Y[,13],2,2),
              substr(Y[,14],1,1),substr(Y[,14],2,2),
              substr(Y[,15],1,1),substr(Y[,15],2,2),
              substr(Y[,16],1,1),substr(Y[,16],2,2),
              substr(Y[,17],1,1),substr(Y[,17],2,2),
              substr(Y[,18],1,1),substr(Y[,18],2,2),
              substr(Y[,19],1,1),substr(Y[,19],2,2),
              substr(Y[,20],1,1),substr(Y[,20],2,2),
              substr(Y[,21],1,1),substr(Y[,21],2,2),
              substr(Y[,22],1,1),substr(Y[,22],2,2),
              substr(Y[,23],1,1),substr(Y[,23],2,2),
              substr(Y[,24],1,1),substr(Y[,24],2,2),
              substr(Y[,25],1,1),substr(Y[,25],2,2),
              substr(Y[,26],1,1),substr(Y[,26],2,2),
              substr(Y[,27],1,1),substr(Y[,27],2,2),
              substr(Y[,28],1,1),substr(Y[,28],2,2))



snpnames <- paste("snp", 1:28,sep="")

out <-  haplo.em(Geno,
                 locus.label=snpnames,
                 control=haplo.em.control(min.posterior=1e-4))

out
```

In total there are 11 haplotypes with "TCACCACTAATGCCAACACGGCCCCACT" being the most popular haplotype with probability of 72.662 %




```{r echo=FALSE, warning=FALSE}


```

```{r echo=FALSE, warning=FALSE}


```


```{r echo=FALSE, warning=FALSE}


```


```{r echo=FALSE, warning=FALSE}


```