---
title: "BSG HW3"
author: "Sten-Oliver Salumaa, Denys Kovalenko"
date: "13 December 2017"
output: html_document
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Task1
#### Myoglobin is an oxygen-binding protein found in muscle tissue. The protein is encoded by the MB gene, which resides on the long arm of chromosome 22. The file MB.rda contains genotype information of unrelated individuals for a set of SNPs in the MB gene. The file contains genotype information in object Y. Load this data into the R environment.

```{r echo=FALSE, warning=FALSE}
load("MB.rda")
```


## Task2
####  (1p) How many individuals and how many SNPs are there in the database? What percentage of the data is missing? 

```{r echo=FALSE, warning=FALSE}
print("Individual count:")
nrow(Y)
print("SNP count:")
ncol(Y)

total <- ncol(Y)*nrow(Y)
nas <- length(Y[is.na(Y)])
print("Percentage of data missing:")
nas/total*100
```


## Task3
#### Assuming all SNPs are bi-allelic, how many haplotypes can theoretically be found for this data set?

```{r echo=FALSE, warning=FALSE}
# not sure but looks like 2**(num of SNPs)
print("Theoretical number of haplotypes:")
2**ncol(Y)
```

## Task4
#### Estimate haplotype frequencies using the haplo.stats package (set the minimum posterior probability to 0.001). How many haplotypes do you find? List the haplotypes and their estimated probabilities. Which haplotype is the most common?

```{r echo=FALSE, warning=FALSE}
library(haplo.stats)

# Remove all SNPs with one or more missing observations.
nmis <- function(x) {
  y <- sum(is.na(x))
}

# get the nmis of SNPs
mispersnp <- apply(Y,1,nmis)
#mispersnp

#dim(Y)
Y2 <- Y[mispersnp==0,]
#dim(Y2)
# no se, we have to produce columns like "A G D C A D S D A A" or whatever and thed we can use the other function
# world-class coding right here:
Geno2 <- cbind(substr(Y2[,1],1,1),substr(Y2[,1],2,2),
              substr(Y2[,2],1,1),substr(Y2[,2],2,2),
              substr(Y2[,3],1,1),substr(Y2[,3],2,2),
              substr(Y2[,4],1,1),substr(Y2[,4],2,2),
              substr(Y2[,5],1,1),substr(Y2[,5],2,2),
              substr(Y2[,6],1,1),substr(Y2[,6],2,2),
              substr(Y2[,7],1,1),substr(Y2[,7],2,2),
              substr(Y2[,8],1,1),substr(Y2[,8],2,2),
              substr(Y2[,9],1,1),substr(Y2[,9],2,2),
              substr(Y2[,10],1,1),substr(Y2[,10],2,2),
              substr(Y2[,11],1,1),substr(Y2[,11],2,2),
              substr(Y2[,12],1,1),substr(Y2[,12],2,2),
              substr(Y2[,13],1,1),substr(Y2[,13],2,2),
              substr(Y2[,14],1,1),substr(Y2[,14],2,2),
              substr(Y2[,15],1,1),substr(Y2[,15],2,2),
              substr(Y2[,16],1,1),substr(Y2[,16],2,2),
              substr(Y2[,17],1,1),substr(Y2[,17],2,2),
              substr(Y2[,18],1,1),substr(Y2[,18],2,2),
              substr(Y2[,19],1,1),substr(Y2[,19],2,2),
              substr(Y2[,20],1,1),substr(Y2[,20],2,2),
              substr(Y2[,21],1,1),substr(Y2[,21],2,2),
              substr(Y2[,22],1,1),substr(Y2[,22],2,2),
              substr(Y2[,23],1,1),substr(Y2[,23],2,2),
              substr(Y2[,24],1,1),substr(Y2[,24],2,2),
              substr(Y2[,25],1,1),substr(Y2[,25],2,2),
              substr(Y2[,26],1,1),substr(Y2[,26],2,2),
              substr(Y2[,27],1,1),substr(Y2[,27],2,2),
              substr(Y2[,28],1,1),substr(Y2[,28],2,2))



snpnames <- paste("snp", 1:28,sep="")

out <-  haplo.em(Geno2,
                 locus.label=snpnames,
                 control=haplo.em.control(min.posterior=1e-4))



print("In total there are 11 haplotypes with")
paste(tail(out$haplotype,1))
print("being the most popular haplotype with probability of")
tail(out$hap.prob,1)

```


## Task5
####  Is the haplotypic constitution of any of the individuals in the database ambiguous or uncertain? If so, for which individuals? What is, in case, the most likely haplotypic constitution of any possibly uncertain individuals?

I have no idea

```{r echo=FALSE, warning=FALSE}



```

```{r echo=FALSE, warning=FALSE}


```

## Task 6. (1p)
#### Suppose we would delete SNP rs5999890 from the database prior to haplotype estimation. Would this affect the results obtained? Justify your answer. Delete this SNP from the database and estimate again the haplotype frequencies. List the haplotypes and their estimated frequencies.

It shouldn't affect results, cause SNP rs5999890 has only CC values, so it doesn't change possible haplotypes.
```{r echo=FALSE, warning=FALSE}
 Y_cut <- Y[,-5]

```


```{r echo=FALSE, warning=FALSE}


```