---
title: "Bioinformatics and Statistical Genetics (thaught by Jan Graffelman), HW3, Linkage Disequilibrium"
author: "Sten-Oliver Salumaa, Denys Kovalenko"
date: "8 December 2017"
output: html_document
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Task1
#### The file ABO-CHB.rda contains genotype information of individuals of a Chinese population of unrelated individuals. The genotype information concerns SNPs the ABO bloodgroup region, located on chromosome number 9. The file contains genotype information (Z, individuals in columns, SNPs in rows), the physical position of each SNP (pos) and the alleles for each SNP (alleles). Load this data into the R environment.

```{r echo=FALSE, message=FALSE, warning=FALSE}
# Load the data
library(genetics)
library(HardyWeinberg)
library(LDheatmap)

filename = "ABO-CHB.rda"
load(filename)
```

Data loaded.

## Task2 (1p)
#### How many individuals and how many SNPs are there in the database?

```{r echo=FALSE}
print("Individual count:")
ncol(Z)
print("SNP count:")
nrow(Z)
```
#### What percentage of the data is missing?

```{r echo=FALSE}
total <- ncol(Z)*nrow(Z)
nas <- length(Z[is.na(Z)])
print("Percentage of data missing:")
nas/total*100
```

## Task3 (1p)
#### Depict all SNPs simultaeneously in a ternary plot, and comment on your result. Do you believe Hardy-Weinberg equilibrium is tenable for the markers in this database?

```{r echo=FALSE}
# transpose the data to make it more logical form -  rows per individuals
Zt <- t(Z)

# create a map function for groupgenotypes
map <- list("AA"=c("A/A"),
            "TT"=c("T/T"),
            "CC"=c("C/C"),
            "GG"=c("G/G"),
            "AB"=".else")

geno_loci <- data.frame(AA = numeric(0), AB = numeric(0), BB = numeric(0))
for(i in 1:ncol(Zt)){
  # genotype counts table as AA, AB, and BB
  #print(expectedGenotypes(genotype(Zt[,i], sep=""), ploidy=2))
  genocounts <- table(groupGenotype(x=genotype(Zt[,i], sep=""), map=map, factor=FALSE))
  #print(genocounts)
  #print(names(genocounts))
  #passing genotype counts to table
  if(length(genocounts)==3){
    geno_loci[i,"AB"] <- genocounts["AB"]  
    genocounts <- genocounts[names(genocounts) != "AB"]
    geno_loci[i, "AA"] <- genocounts[1]
    geno_loci[i, "BB"] <- genocounts[2]
  }else if(length(genocounts)==2){
    geno_loci[i,"AB"] <- genocounts["AB"]  
    genocounts <- genocounts[names(genocounts) != "AB"]
    geno_loci[i, "AA"] <- genocounts[1]
  }else if(length(genocounts)==1){
    geno_loci[i,"AB"] <- genocounts["AB"]  
    genocounts <- genocounts[names(genocounts) != "AB"]
    geno_loci[i, "AA"] <- genocounts[1]
  }

}

# replace n.a.-s
geno_loci[is.na(geno_loci)] <- 0
# plot the thing
HWTernaryPlot(geno_loci)
```

Yes, it seems that HW equilibrium is tenable for this database since almost all of the data points reside in the equilibrium boundaries.


## Task4 (1p)

#### Using the function LD from the genetics package, compute the LD statistic D for the first two SNPs in the database. Is there significant association between these two SNPs?

```{r echo=FALSE}

SNP1g <- genotype(Zt[,1],sep="")
SNP2g <- genotype(Zt[,2],sep="")

res <- LD(SNP1g,SNP2g)
res
print("P-value:")
res$`P-value`

```
D' is close to 1, so first assumption is that we might have high LD.
But we think there is no significant association between these two SNPs because the p-value is too high (above our threshold of 5 %)



## Task5 (2p)
#### Given your previous estimate of D, and using the formulae from the lecture slides, compute the statistics D', Chi2, R2 and r by hand for the first pair of SNPs. Do your results coincide with those obtained by the LD function? Can you explain possible differences?

```{r echo=FALSE}
#print(geno_loci[1:2,])

#nice example http://pbgworks.org/sites/pbgworks.org/files/measuresoflinkagedisequilibrium-111119214123-phpapp01_0.pdf
#table(SNP1g)
#table(SNP2g)

#D = p(AB) - pA*pB -> A = G in 1st SNP, a = A in 1st SNP, B = G in 2nd SNP, b = A in 2nd SNP
#Dcalc = 68/90 - (157/180)*(157/180)# -> 68/90
Dcalc = res$D
print("D:")
Dcalc

pApB=(157/180)*(157/180)
papb <- 23/180*23/180
Dmax <- min(pApB, papb)

print("D':")
D_p <- Dcalc/Dmax
D_p
# -0.3194707, doesn't show high LD...

r2 <- Dcalc^2 / (pApB*papb)
print("r**2: ")
r2

# X^2
X2 <- r2 * 2*nrow(Zt)
print("X**2: ")
X2
```
Differences for D occur because we used a small sample for our calculations (we read that with small sample size D' can be inflated).



## Task 6 (2p)
####  Given your previous estimate of D, infer the haplotype frequencies. Which haplotype is the most common?

```{r echo=FALSE}
#P11 = P(GG) = p(G)*p(G) + D
#P22 = P(AA) = p(A)*p(A) + D
#P12 = P(GA) = p(G)*p(A) - D
#P21 = P(AG) = p(A)*p(G) - D

D <- Dcalc

P11 <- (157/180)**2 + D
P22 <- (23/180)**2 + D
P12 <- (157/180)*(23/180) - D
P21 <-(157/180)*(23/180) - D

max(P11, P22, P12, P21)

```
Most common is GG haplotype with probability of 0.744763.




## Task 7 (2p)
#### Compute 4 LD statistics for all the marker pairs in this data base (D, D', �2 and R2). Make a scatterplot matrix of these. Is there an exact linear relationship between �2 and R2 ? Why (not) so?

```{r echo=FALSE}
p<-NULL
l <- ncol(Zt)-1
c = matrix(data = 0, nrow=nrow(Zt), ncol=ncol(Zt))
for(i in 1:l) {
  for(j in i:ncol(Zt)) {
    SNP1gTmp <- genotype(Zt[,i],sep="")
    SNP2gTmp <- genotype(Zt[,j],sep="")
    resTmp <- LD(SNP1gTmp,SNP2gTmp)
    p$D[c] <- resTmp$D
    p$`D'`[c] <- resTmp$`D'`
    p$`R^2`[c] <- resTmp$`R^2`
    p$`X^2`[c] <- resTmp$`X^2`
    c = c+1
  }
}

pairs(p)
```

There is linear relationship between X^2 and R^2 (line in form y = multiplier*x). By definition, X2 and R2 are tied by equation R2=X2/2n (number of chromosomes)

## Task 8. (2p)
#### Compute a distance matrix with the distance in base pairs between all possible pairs of SNPs. Make a plot of the R2 statistics against the distance between the markers. Comment on your results

```{r echo=FALSE}
distance <- function(x,y) {
  if (is.na(x) || is.na(y)) {
    return (0)
  }
  ly <- nchar(y)
  lx <- nchar(x)
  min <- min(ly,lx)
  delta = 0
  for (i in 1:min) {
    if (substr(x, i,i)!=substr(y, i, i)) {
      delta = delta + 1
    }
  }
  return(delta)
}

snpDistance <- function(snp1, snp2) {
  d = 0
  for (i in 1:length(snp1)) {
    d = d + distance(snp1[i], snp2[i])
  }
  return (d)
}
distance_matrix <- matrix(data = NA, nrow = nrow(Zt), ncol= ncol(Zt))
r2_matrix <- matrix(data = 0, nrow = nrow(Zt), ncol= ncol(Zt))
for(i in 1:ncol(Zt)) {
  for(j in 1:ncol(Zt)) {
    distance_matrix[i,j] = snpDistance(Zt[,i], Zt[,j])
    r2_matrix[i,j] <- LD(genotype(Zt[,i],sep=""),genotype(Zt[,j],sep=""))$`R^2`
  }
}
r2 <- r2_matrix[upper.tri(r2_matrix)]
dist <- distance_matrix[upper.tri(distance_matrix)]

plot(dist, r2, ylab="R^2 statistic", xlab = "distance between SNPs")

```

There is no clear trend (correlation) between distance between SNPs char-wise and R^2. Which means that having very similar SNPs doesn't affect LD between them.


## Task 9. (2p) 
#### Make two LD heatmaps of the markers in this database, one using the R2 statistic and one using the D' statistic, and use the positional information on the markers. Are the results consistent? 

```{r echo=FALSE}
#### ... we don't have SNPs labels here as it was in lecture.
RES <- data.frame(genotype(Zt[,1],sep=""))

for(i in 2:ncol(Zt)) {
   snp <- genotype(Zt[,i],sep="")
   RES <- cbind(RES,snp)
}
colnames(RES) <- 1:45
output <- LD(RES)

output <- LD(RES)

Dm <- output$D
Dp <- output$"D'"
R2 <- output$"R^2"
X2 <- output$"X^2"

#Dm[1:10,1:10]

Dm <- Dm[upper.tri(Dm)]
Dp <- Dm[upper.tri(Dp)]
R2 <- Dm[upper.tri(R2)]
X2 <- Dm[upper.tri(X2)]

#Dm
#pairs(cbind(Dm,Dp,R2,X2))
#
#dim(output)
rgb.palette <- colorRampPalette(rev(c("blue", "orange", "red")), space = "rgb")
LDheatmap(output$`D'`, genetic.distances=pos, LDmeasure="D'", add.map=TRUE, title="D' LD heatmap",color=rgb.palette(18))

```

```{r echo=FALSE}
LDheatmap(output$`R^2`,LDmeasure="r", genetic.distances=pos, add.map=TRUE, title="R^2 LDheatmap",color=rgb.palette(18))
```

Results are incosistent - with D' LDHeatMap we can see that D's are uniformly distributed and are close to 1 for most of SNP pairs, while with R^2 we see several clusters of closely positioned SNPS to have high R^2


## Task 10. (2p) 

#### Simulate 45 independent SNPs under the assumption of Hardy-Weinberg equilibrium. Simulate as many SNPs as you have in your database, and take care to match each SNP in your database with a simulated SNP that has the same sample size and allele frequency. Make two LD heatmaps of the simulated SNPs, one using R^2 and one using D'. Compare these to the LD heatmap of the ABO region. What do you observe? State your conclusions


```{r echo=FALSE, warning=FALSE}

HWsimulated <- HWData(45,45)

allelef <- numeric(45)
for(i in 1:dim(geno_loci)[1]){
  aa <-geno_loci[i, 1]
  ab <-geno_loci[i,2]
  bb <-geno_loci[i,3]
  allelef[i] = (aa+ab/2)/(aa+ab+bb)
}


simulated <- HWData(nm = 45, n = 45, p=allelef)

sim_df <- data.frame(row.names = 1:45)
for(i in 1:nrow(simulated)){
    generated_row = c()
  row <- simulated[i,]
  for(j in 1:length(row)){
    
    snp <- colnames(simulated)[j]
    reps <- rep(snp, row[j])
    generated_row <- c(generated_row, reps)
    #append(generated_row, reps)
    #append(x,reps)
    #print(generated_row)
    
  }
  sim_df <- cbind(sim_df, data.frame(genotype(generated_row, sep = "")))
}


RES3 <- data.frame(genotype(sim_df[,1]))

for(i in 2:ncol(sim_df)) {
   snp <- genotype(sim_df[,i])
   RES3 <- cbind(RES3,snp)
}
colnames(RES3) <- 1:45
RES3 <- RES3[sample(nrow(RES3)),]
rownames(RES3) <- 1:45
output <- LD(RES3)


rgb.palette <- colorRampPalette(rev(c("blue", "orange", "red")), space = "rgb")
LDheatmap(output$`R^2`,LDmeasure="r", genetic.distances=pos, add.map=TRUE, title="R^2 LDheatmap",color=rgb.palette(18))
LDheatmap(output$`D'`, genetic.distances=pos, LDmeasure="D'", add.map=TRUE, title="D' LD heatmap",color=rgb.palette(18))
```

The first LD heatmap using measure of D' look similar to the AB0 region data. The differences step in on the second heatmap using R^2 data. The AB0 region SNPs seem to have clusters on the heatmap - regions that clearly influence each other. But generated data doesn't show any clear regions.

## Task 11 (1p) 
####Do you think there is strong or weak LD for the ABO region you just studied? Explain your opinion

There is a strong LD for the AB0 region. We see high D' values which means that (almost) all SNPs of this region are strongly associated. Also, we see some low R^2 values which doesn't directly mean that there is weak LD but might also mean that in some loci - although having strong association - are represented with low quantities. But we can see regions even in R^2 LD heatmap that have high values and thus showing high LD (same region on D' heatmap is high-valued as well).

