---
title: "Bioinformatics and Statistical Genetics (taught by Jan Graffelman), HW5, Population Substructure"
author: "Sten-Oliver Salumaa, Denys Kovalenko"
date: "11 January 2017"
output: html_document
---


```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(genetics)
```

## Task1
#### The file SNPChr20.rda contains genotype information of 310 individuals of unknown background. The genotype information concerns 50.000 SNPs on chromosome 20. Load this data into the R environment. The data file contains a matrix Y containing the allele counts (0,1 or 2) for 50.000 SNPs for one of the alleles of each SNP

```{r echo=FALSE, warning=FALSE}
load("SNPChr20.rda")
```


## Task 2
#### Compute the Manhattan distance matrix between the 310 individuals (this may take a few minutes) Include a submatrix of dimension 5 by 5 with the distances between the first 5 individuals in your report

```{r echo=FALSE, warning=FALSE}
#dim(Y)
#Y[1:5, 1:5]
dist_Y_man <- dist(Y, method="manhattan", diag = FALSE)
D <- as.matrix(dist_Y_man)
#following the definition above
#manhattan_distance <- function(p,q){
#  sum(abs(p-q))
#}
#manhattan distance from 6 to 7
#manhattan_distance(as.numeric(Y[1,]),as.numeric(Y[2,]))
#sum(abs(as.numeric(Y[1,]) - as.numeric(Y[2,])), na.rm = TRUE)
print("First 5 rows and columns of manhattan distance calculations: ")
D[1:5, 1:5]

```

## Task 3
#### Use metric multidimensional scaling to obtain a map of the individuals, and include your map in your report. Do you think the data come from one homogeneous population?

```{r echo=FALSE, warning=FALSE}
dist_Y_euc <- dist(Y)
D2 <- as.matrix(dist_Y_euc) # why using euclidean here?

n <- nrow(Y)
mds.out <- cmdscale(D2, k=n-1, eig=TRUE)
X <- mds.out$points[,1:2]
plot(X[,1],X[,2], asp=1, xlab = "First principal axis", ylab="Second principal axis")
```

This map clearly shows that there are 3 different population groups involved. This data did not come from a homogenous population.


## Task 4
####  Report the eigenvalues of the solution

```{r echo=FALSE, warning=FALSE}
mds.out$eig
```

## Task 5
####  Is the distance matrix you have used an Euclidean distance matrix?

For the values in Task 2 we used Manhattan distance and for the calculations to plot the multidimensional scaling results we used euclidian distance.
Also, the function 'is.euclid' from library 'ade4' confirms our answer.

```{r warning=FALSE}
library(ade4)
D2 <- dist_Y_euc
is.euclid(D2)
```


## Task 6
#### What is the goodness-of-fit of a two-dimensional approximation to your distance matrix?
Here we are applying the same 'cmdscale' function to the previously calculated eucledian distances but this time with 'k' value as 2. Reporting the GOF:

```{r echo=FALSE, warning=FALSE}
mds.out.2d <- cmdscale(D2, k=2, eig=TRUE)
mds.out.2d$GOF
```

## Task 7
#### Make a plot of the estimated distances (according to your map of individuals) versus the observed distances.

```{r echo=FALSE, warning=FALSE}
Dest <- as.matrix(dist(X)) # creates all distances between points I have on my computer screen
#is.euclid(dist(X)) # TRUE
#is.euclid(dist_Y_euc) # TRUE

Dobs.vec <- D2[lower.tri(D2)] #gives an indicator matrix whether a particular entry is below diagonal or no
Dest.vec <- Dest[lower.tri(Dest)] # returns values only below the diagonal


plot(Dobs.vec, Dest.vec, xlab="Observed", ylab="fitted")

```

#### Regress estimated distances on observed distances and report the coefficient of determination of the regression

```{r echo=FALSE, warning=FALSE}
# regressing
regr <- lm(Dobs.vec ~ Dest.vec)
print("R-squared value: ")
summary(regr)$r.squared
```

## Task 8
#### Try now non-metric multidimensional scaling with your distance matrix. Use both a random initial configuration as well as the classical metric solution as an initial solution. Make a plot of the two-dimensional solution. Do the results support that the data come from one homogeneous population?

```{r echo=FALSE, warning=FALSE}
library(MASS)
?isoMDS
set.seed(105)
yinit <- scale(matrix(runif(2*n),ncol=2),scale = FALSE)
nmds.out_r <- isoMDS(D,k=2,y=yinit)
plot(nmds.out_r$points[,1], nmds.out_r$points[,2], xlab = "1st PC", ylab = "2nd PC")

```


```{r echo=FALSE, warning=FALSE}
nmds.out_cl <- isoMDS(D, k=2)
plot(nmds.out_cl$points[,1], nmds.out_cl$points[,2], xlab = "1st PC", ylab = "2nd PC", main="Fitted configuration with classical solution as init conf")
```

## No, we got to see 3 clusters with NMDS for both classical and and random initial solutions, which supports that data comes not from one homogeneous population.


## Task 9.
#### (1p) Make again a plot of the estimated distances (according to your map of individuals) versus the observed distances, now for the two-dimensional solution of non-metric MDS. Regress estimated distances on observed distances and report the coefficient of determination of the regression. Is the fit better or worse than with metric MDS?

```{r echo=FALSE, warning=FALSE}
Dest_nm <- as.matrix(dist(nmds.out_cl$points[,1:2]))
Dest_nm.vec <- Dest_nm[lower.tri(Dest_nm)] 
plot(Dobs.vec, Dest_nm.vec, xlab="Observed", ylab="fitted")
```
```{r echo=FALSE, warning=FALSE}
# regressing
regr <- lm(Dobs.vec ~ Dest_nm.vec)
print("R-squared value: ")
summary(regr)$r.squared
```
Plots look the same, but R^2 is lower for Non-metric scaling, which ??????????????????????

## Task 10(1p) Compute the stress for a 1, 2, 3, 4, . . . , n-dimensional solution, always using the classical MDS solution as an initial configuration. How many dimensions are necessary to obtain a good representation? Make a plot of the stress against the number of dimensions
```{r echo=FALSE, warning=FALSE}
dimensions = seq(1, 40)
stress_l <- NULL
set.seed(105)
for(i in 1:length(dimensions)) {
  stress_l[i] <- isoMDS(D,k=i)$stress
}
plot(dimensions, stress_l, xlab="dimensions", ylab="stress", main="Stress over dimensions")
```
After 15, stress decreases more slowly. Altough on observable dimensions rate of decrease stays the same value of stress itself is pretty low at 30 already.



## Task 11. 
#### Compute the correlation matrix between the first two dimensions of a metric MDS and the two-dimensional solution of a non-metric MDS. Make a scatterplot matrix of the 4 variables. Comment on your findings 
```{r echo=FALSE, warning=FALSE}
p <- NULL
p$pc1m = X[,1]
p$pc2m = X[,2]
p$pc1nm = nmds.out_cl$points[,1]
p$pc2nm = nmds.out_cl$points[,2]

corr_matrix = matrix(data = 0, nrow=2, ncol=2)
corr_matrix[1,1] = cor(p$pc1m, p$pc1nm)
corr_matrix[1,2] = cor(p$pc1m, p$pc2nm)
corr_matrix[2,1] = cor(p$pc2m, p$pc1nm)
corr_matrix[2,2] = cor(p$pc2m, p$pc2nm)



pairs(p, main="scatterolot matrix")
```


```{r echo=FALSE, warning=FALSE}
corr_matrix
```


```{r echo=FALSE, warning=FALSE}

```


```{r echo=FALSE, warning=FALSE}

```

